### ISSUE TITLE
Qt + cmake-conan + Qt Creator + Ubuntu = failure to install

### ISSUE THREAD

**Issue by ArekPiekarz** [bincrafters/community#378]
### Description of Problem

Qt package used with the cmake-conan wrapper in Qt Creator in Ubuntu fails on `conan install` step. The logs are sparse (because of bug in CMake https://gitlab.kitware.com/cmake/cmake/issues/17931), but my theory is it cannot receive password for `sudo apt update` because of noninteractive console.

### Package Details
* Package Name/Version: **Qt/5.11.1**
* Operating System: **Ubuntu 18.04 x64**
* Compiler+version: **GCC 8.1**
* CMake: **3.12**
* cmake-conan: **0.12**
* Qt Creator: **4.7.0**

### Steps to reproduce
conanfile.txt
```
[requires]
Qt/5.11.1@bincrafters/stable

[options]
*:shared=True

[generators]
cmake
```
CMakeLists.txt
```
cmake_minimum_required(VERSION 3.10)

project(conan-qt-test LANGUAGES CXX)
set(CMAKE_CXX_STANDARD 17)
set(CMAKE_CXX_STANDARD_REQUIRED ON)
set(CMAKE_CXX_EXTENSIONS OFF)

if(NOT EXISTS "${CMAKE_BINARY_DIR}/conan.cmake")
   message(STATUS "Downloading conan.cmake from https://github.com/conan-io/cmake-conan")
   file(DOWNLOAD "https://raw.githubusercontent.com/conan-io/cmake-conan/v0.12/conan.cmake"
       "${CMAKE_BINARY_DIR}/conan.cmake")
endif()
include(${CMAKE_BINARY_DIR}/conan.cmake)
conan_cmake_run(
    CONANFILE conanfile.txt
    BASIC_SETUP
    BUILD missing)

add_executable(${PROJECT_NAME} main.cpp)
target_link_libraries(${PROJECT_NAME} ${CONAN_LIBS})
```
main.cpp
```
#include <QtCore/QDebug>

int main()
{
    qDebug() << "hello\n";
    return 0;
}
```
**Steps:**
* Open Qt Creator
* Click File -> Open file or project
* Find CMakeLists.txt and click Open
* Select your kit, click on details, uncheck Default, click Configure Project

**Expected result:**
CMake + Conan configuration should succeed.

**Actual result:**
CMake + Conan configuration fails with:
```
Running "/usr/local/bin/cmake -E server --pipe=/tmp/cmake-.BlsLST/socket --experimental" in /home/fazer/dev/project/test/build-conan-cmake-qtcreator-bug-Desktop-Debug.
Starting to parse CMake project, using: "-DCMAKE_BUILD_TYPE:STRING=Debug", "-DCMAKE_CXX_COMPILER:STRING=/usr/bin/g++-8", "-DCMAKE_C_COMPILER:STRING=/usr/bin/gcc-8", "-DCMAKE_PREFIX_PATH:STRING=".
The CXX compiler identification is GNU 8.1.0
Check for working CXX compiler: /usr/bin/g++-8
Check for working CXX compiler: /usr/bin/g++-8 -- works
Detecting CXX compiler ABI info
Detecting CXX compiler ABI info - done
Detecting CXX compile features
Detecting CXX compile features - done
Downloading conan.cmake from https://github.com/conan-io/cmake-conan
Conan ** WARNING** : This detection of settings from cmake is experimental and incomplete. Please check 'conan.cmake' and contribute
Conan executing: conan install /home/fazer/dev/project/test/conan-cmake-qtcreator-bug/conanfile.txt -g cmake -s build_type=Debug -s os=Linux -s compiler=gcc -s compiler.version=8 -s compiler.libcxx=libstdc++11 --build=missing
CMake Error at /home/fazer/dev/project/test/build-conan-cmake-qtcreator-bug-Desktop-Debug/conan.cmake:335 (message):
  Conan install failed='1'
Call Stack (most recent call first):
  /home/fazer/dev/project/test/build-conan-cmake-qtcreator-bug-Desktop-Debug/conan.cmake:413 (conan_cmake_install)
  CMakeLists.txt:14 (conan_cmake_run)


Configuring incomplete, errors occurred!
See also "/home/fazer/dev/project/test/build-conan-cmake-qtcreator-bug-Desktop-Debug/CMakeFiles/CMakeOutput.log".
CMake Project parsing failed.
```

**Logs:**
[build-conan-cmake-qtcreator-bug-Desktop-Debug.tar.gz](https://github.com/bincrafters/community/files/2216057/build-conan-cmake-qtcreator-bug-Desktop-Debug.tar.gz)

**Comment by ericLemanissier** [comment]
I made a test with the example you give, but starting the build via ` cmake . -DCMAKE_BUILD_TYPE=Release`
System packages have been properly installed, after typing the password:
```
-- The CXX compiler identification is GNU 5.4.0
-- Check for working CXX compiler: /usr/bin/c++
-- Check for working CXX compiler: /usr/bin/c++ -- works
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Conan ** WARNING** : This detection of settings from cmake is experimental and incomplete. Please check 'conan.cmake' and contribute
-- Conan executing: conan install /home/eric/conantest/conanfile.txt -g cmake -s build_type=Release -s os=Linux -s compiler=gcc -s compiler.version=5 -s compiler.libcxx=libstdc++11 --build=missing
Running: sudo apt-get update
[sudo] password for eric:
Hit:1 http://archive.ubuntu.com/ubuntu xenial InRelease
Get:2 http://archive.ubuntu.com/ubuntu xenial-updates InRelease [109 kB]
Get:3 http://archive.ubuntu.com/ubuntu xenial-backports InRelease [107 kB]
Get:4 http://security.ubuntu.com/ubuntu xenial-security InRelease [107 kB]
Fetched 323 kB in 1s (300 kB/s)
Reading package lists...
dpkg-query: package 'libxcb1' is not installed and no information is available
(...)
```
I agree with your theory that the non interactive console prevents the installation of system packages if password is needed.
Could you try calling cmake manually from an interactive console ? Once the system packages are installed, it should work also from qtcreator. If it solves the problem, I think the issue should be reported to conan-io/cmake-conan

**Comment by ArekPiekarz** [comment]
Unfortunately, that doesn't solve the problem. Here's what I did:
* mkdir build-release && cd build-release
* cmake .. -DCMAKE_BUILD_TYPE=Release
* Open Qt Creator
* Click File -> Open file or project
* Find CMakeLists.txt and click Open
* Uncheck Desktop kit (my default one)
* Leave the first "Imported kit - temporary" kit that has Release checked
* Uncheck the second "Imported kit -temporary" kit that has "Debug" checked
* Click Configure Project

After that I get the same error message as earlier, so cmake-conan still tries to run `conan install`.

According to your logic in " If it solves the problem, I think the issue should be reported to conan-io/cmake-conan", I should not report it to them, because the issue is not solved. Is that what you intended? :-)

**Comment by ericLemanissier** [comment]
Did you try to remove Qt from conanfile.txt (and remove qdebug from main.cpp) ?
if it fails too, then please report to cmake-conan.
If it succeeds, 
1. put qt back in conanfile.txt
2. retry building from qt creator
3. check in the binary dir if anything was installed
4. try running `conan install /home/fazer/dev/project/test/conan-cmake-qtcreator-bug/conanfile.txt -g cmake -s build_type=Debug -s os=Linux -s compiler=gcc -s compiler.version=8 -s compiler.libcxx=libstdc++11 --build=missing` from the binary dir, and see what fails

Another idea, did you make sure that the path to conan is in the Path variable in the compilation environment of the project in qt creator ?

**Comment by ArekPiekarz** [comment]
> Did you try to remove Qt from conanfile.txt (and remove qdebug from main.cpp) ?
> if it fails too, then please report to cmake-conan.

The CMake configuration (and later app building) succeeded with a following log, where we can see that `conan` is in the path:
```
Running "/usr/local/bin/cmake -E server --pipe=/tmp/cmake-.iLmNTS/socket --experimental" in /home/fazer/dev/project/test/build-conan-cmake-qtcreator-bug-Desktop-Debug.
Starting to parse CMake project.
Conan ** WARNING** : This detection of settings from cmake is experimental and incomplete. Please check 'conan.cmake' and contribute
Conan executing: conan install /home/fazer/dev/project/test/conan-cmake-qtcreator-bug/conanfile.txt -g cmake -s build_type=Debug -s os=Linux -s compiler=gcc -s compiler.version=8 -s compiler.libcxx=libstdc++11 --build=missing
Conan: Loading conanbuildinfo.cmake
Current conanbuildinfo.cmake directory: /home/fazer/dev/project/test/build-conan-cmake-qtcreator-bug-Desktop-Debug
Conan: Compiler GCC>=5, checking major version 8
Conan: Checking correct version: 8
Conan: Using cmake global configuration
Conan: Adjusting default RPATHs Conan policies
Conan: Adjusting language standard
Configuring done
Generating done
CMake Project was parsed successfully.
```

> If it succeeds,
> put qt back in conanfile.txt
> retry building from qt creator

I did that and finally got a confirmation what causes the problem:
```
13:20:15: Running steps for project conan-qt-test...
13:20:15: Starting: "/usr/local/bin/cmake" --build . --target all
[0/1 ?/sec] Re-running CMake...
-- Conan ** WARNING** : This detection of settings from cmake is experimental and incomplete. Please check 'conan.cmake' and contribute
-- Conan executing: conan install /home/fazer/dev/project/test/conan-cmake-qtcreator-bug/conanfile.txt -g cmake -s build_type=Debug -s os=Linux -s compiler=gcc -s compiler.version=8 -s compiler.libcxx=libstdc++11 --build=missing
Running: sudo apt-get update
sudo: no tty present and no askpass program specified
ERROR: Qt/5.11.1@bincrafters/stable: Error in requirements() method, line 111
	installer.update() # Update the package database
	ConanException: Command 'sudo apt-get update' failed
CMake Error at /home/fazer/dev/project/test/build-conan-cmake-qtcreator-bug-Desktop-Debug/conan.cmake:335 (message):
  Conan install failed='1'
Call Stack (most recent call first):
  /home/fazer/dev/project/test/build-conan-cmake-qtcreator-bug-Desktop-Debug/conan.cmake:413 (conan_cmake_install)
  CMakeLists.txt:14 (conan_cmake_run)


-- Configuring incomplete, errors occurred!
See also "/home/fazer/dev/project/test/build-conan-cmake-qtcreator-bug-Desktop-Debug/CMakeFiles/CMakeOutput.log".
FAILED: build.ninja 
/home/fazer/dev/app/cmake/bin/cmake -H/home/fazer/dev/project/test/conan-cmake-qtcreator-bug -B/home/fazer/dev/project/test/build-conan-cmake-qtcreator-bug-Desktop-Debug
ninja: error: rebuilding 'build.ninja': subcommand failed
13:20:15: The process "/usr/local/bin/cmake" exited with code 1.
Error while building/deploying project conan-qt-test (kit: Desktop)
When executing step "CMake Build"
13:20:15: Elapsed time: 00:00.
```
I took a liberty to skip the rest of proposed steps, I hope you don't mind. I'll try to explore if Qt Creator can somehowe work with askpass, unless you have a better idea (like a configuration option to skip invoking apt-get).

**Comment by ericLemanissier** [comment]
I may have a solution which does not invoke the system package manager if the package is already built. It would require you to build the package from TTY beforehand at least once with the same options/settings though.
Could you try with https://github.com/ericLemanissier/conan-qt/tree/stable/5.11.1 ?

**Comment by ArekPiekarz** [comment]
I configured CMake in console, then launched the project in Qt Creator and the configuration succeeded.
The building of application then failed on:
```
16:44:08: Running steps for project conan-qt-test...
16:44:08: Starting: "/usr/local/bin/cmake" --build . --target all
Scanning dependencies of target conan-qt-test
[ 50%] Building CXX object CMakeFiles/conan-qt-test.dir/main.cpp.o
In file included from /home/fazer/.conan/data/Qt/5.11.1/ericlemanissier/stable/package/84831faa6790e9dd8685247518b48aa254ec78b7/include/QtCore/qalgorithms.h:43,
                 from /home/fazer/.conan/data/Qt/5.11.1/ericlemanissier/stable/package/84831faa6790e9dd8685247518b48aa254ec78b7/include/QtCore/qdebug.h:44,
                 from /home/fazer/.conan/data/Qt/5.11.1/ericlemanissier/stable/package/84831faa6790e9dd8685247518b48aa254ec78b7/include/QtCore/QDebug:1,
                 from /home/fazer/dev/project/test/conan-cmake-qtcreator-bug/main.cpp:1:
/home/fazer/.conan/data/Qt/5.11.1/ericlemanissier/stable/package/84831faa6790e9dd8685247518b48aa254ec78b7/include/QtCore/qglobal.h:1177:4: error: #error "You must build your code with position independent code if Qt was built with -reduce-relocations. " "Compile your code with -fPIC (-fPIE is not enough)."
 #  error "You must build your code with position independent code if Qt was built with -reduce-relocations. "\
    ^~~~~
CMakeFiles/conan-qt-test.dir/build.make:62: recipe for target 'CMakeFiles/conan-qt-test.dir/main.cpp.o' failed
make[2]: *** [CMakeFiles/conan-qt-test.dir/main.cpp.o] Error 1
make[1]: *** [CMakeFiles/conan-qt-test.dir/all] Error 2
make: *** [all] Error 2
CMakeFiles/Makefile2:72: recipe for target 'CMakeFiles/conan-qt-test.dir/all' failed
Makefile:83: recipe for target 'all' failed
16:44:09: The process "/usr/local/bin/cmake" exited with code 2.
Error while building/deploying project conan-qt-test (kit: Imported Kit)
When executing step "CMake Build"
16:44:09: Elapsed time: 00:01.
```
Weren't there some changes in Conan to handle it automatically? Anyway, I added:
`add_compile_options(-fPIC)`
before `if(NOT EXISTS "${CMAKE_BINARY_DIR}/conan.cmake")`
and then got:
```
16:49:19: Running steps for project conan-qt-test...
16:49:19: Starting: "/usr/local/bin/cmake" --build . --target all
[ 50%] Building CXX object CMakeFiles/conan-qt-test.dir/main.cpp.o
[100%] Linking CXX executable bin/conan-qt-test
CMakeFiles/conan-qt-test.dir/main.cpp.o:(.qtversion[qt_version_tag]+0x0): undefined reference to `qt_version_tag'
CMakeFiles/conan-qt-test.dir/main.cpp.o: In function `main.cold.0':
main.cpp:(.text.unlikely+0x16): undefined reference to `QArrayData::deallocate(QArrayData*, unsigned long, unsigned long)'
CMakeFiles/conan-qt-test.dir/build.make:83: recipe for target 'bin/conan-qt-test' failed
CMakeFiles/Makefile2:72: recipe for target 'CMakeFiles/conan-qt-test.dir/all' failed
Makefile:83: recipe for target 'all' failed
main.cpp:(.text.unlikely+0x1e): undefined reference to `QDebug::~QDebug()'
CMakeFiles/conan-qt-test.dir/main.cpp.o: In function `main':
main.cpp:(.text.startup+0x6e): undefined reference to `QMessageLogger::debug() const'
main.cpp:(.text.startup+0x87): undefined reference to `QString::fromUtf8_helper(char const*, int)'
main.cpp:(.text.startup+0x92): undefined reference to `QTextStream::operator<<(QString const&)'
main.cpp:(.text.startup+0xbb): undefined reference to `QDebug::~QDebug()'
main.cpp:(.text.startup+0xea): undefined reference to `QArrayData::deallocate(QArrayData*, unsigned long, unsigned long)'
main.cpp:(.text.startup+0xf6): undefined reference to `QTextStream::operator<<(char)'
collect2: error: ld returned 1 exit status
make[2]: *** [bin/conan-qt-test] Error 1
make[1]: *** [CMakeFiles/conan-qt-test.dir/all] Error 2
make: *** [all] Error 2
16:49:20: The process "/usr/local/bin/cmake" exited with code 2.
Error while building/deploying project conan-qt-test (kit: Imported Kit)
When executing step "CMake Build"
16:49:20: Elapsed time: 00:01.
```
Later I'll try rebuilding it from scratch.

**Comment by ArekPiekarz** [comment]
I managed to build it only after adding
`find_package(Qt5 5.11 COMPONENTS Core REQUIRED)`
and changing linking to
`target_link_libraries(${PROJECT_NAME} Qt5::Core)`
I also tried to print the content of ${CONAN_LIBS}, but it was empty. Using CONAN_PKG::Qt for linking  didn't work too. 

But the linking seems to be a different problem than the original in this report. For now we could go ahead with your change, Eric.

**Comment by ericLemanissier** [comment]
yes, using cmake with Qt requires the usage of find_package and target_link_libraries(... Qt5::module). This problem is independent from conan and the qt recipe.

**Comment by ericLemanissier** [comment]
this should be partially solved in stable/5.11.1: sudo is not called if packages are already installed.

**Comment by ArekPiekarz** [comment]
I can confirm Qt Creator can now configure and build the project when the Qt package was built beforehand. For me this could be closed and a more robust solution should be investigated in https://github.com/conan-io/conan/issues/3285. If you think otherwise, feel free to reopen. Thanks!
